### PR TITLE
#21772 Flush old host cache when the VanityURL is Updated

### LINKED ISSUE #21772: Unable to Change Hosts for Vanity URLS - 5.3.8 and possibily 21.06
Copied from: #21373 

Found when working on Support Ticket: [105909](https://dotcms.zendesk.com/agent/tickets/105909)
Reproducible as of dotCMS 5.3.8.6 up to latest master (21.11)

Vanity URLs can work for either a specific Site or all Sites in a dotCMS instance. If you change where a given Vanity URL is supposed to work on, it should work as expected. However, when doing so, changes are not getting picked up and the Vanity URL still points to the previous  value -- either a specific Site, or all Sites.

Flushing the Cache Regions and/or doing a full reindex is not fixing the problem.

### NOTE TO DEVS #2: At least in dotCMS 5.3.8.6, adding the `Site or Folder` field does not solve the problem. The back-end is ignoring any change you make to it and the new Site you select is not persisted.

**To Reproduce**
1. Create a copy of the `demo.dotcms.com` Site named `second.dotcms.com`.
2. Select the `demo.dotcms.com` Site, and create a Vanity URL with the following information. **It's very important the the Vanity points to `All sites` first**:
2.1. Title: `My test Vanity`
2.2. Site: `All sites`
2.3. Uri: `/my-test`
2.4. Action: `301 - Redirect`
2.5. Forward To: `/index`
3. The Vanity URL works for `demo.dotcms.com`:
```
$ curl -sk --head -H "Host: demo.dotcms.com"  https://demo.dotcms.com/my-test
HTTP/1.1 301 
Strict-Transport-Security: max-age=3600;includeSubDomains
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
X-DOT-VanityUrl: 315b78b05326fa2033219cc150da4848
Location: /index
Transfer-Encoding: chunked
Date: Mon, 06 Dec 2021 18:40:42 GMT
Connection: close
```
4. And for `second.dotcms.com` as well, which is expected:
```
$ curl -sk --head -H "Host: second.dotcms.com"  https://demo.dotcms.com/my-test
HTTP/1.1 301 
Strict-Transport-Security: max-age=3600;includeSubDomains
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
X-DOT-VanityUrl: 315b78b05326fa2033219cc150da4848
Location: /index
Transfer-Encoding: chunked
Date: Mon, 06 Dec 2021 18:40:58 GMT
Connection: close
```
5. Now, change the Vanity URL to point to `demo.dotcms.com` **ONLY**, and try both CURL commands again.
6. Both of them are returning Status Code 301, which is **NOT** correct as the one sent to `second.dotcms.com` should have failed with a 404.
7. If you re-create the Vanity URL and point it to the `demo.dotcms.com` Site since the beginning, the CURL command to `demo.dotcms.com` works again, and the one to `second.dotcms.com` fails as expected:
```
$ curl -sk --head -H "Host: second.dotcms.com"  https://demo.dotcms.com/my-test
HTTP/1.1 404 
Strict-Transport-Security: max-age=3600;includeSubDomains
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
X-XSS-Protection: 1; mode=block
Set-Cookie: JSESSIONID=D6BA571A1954E4ABD0E9A5951245AEFE; Path=/; Secure; HttpOnly; SameSite=Strict
Content-Type: text/html;charset=ISO-8859-1
Transfer-Encoding: chunked
Date: Mon, 06 Dec 2021 18:45:09 GMT
Connection: close
```

**Expected behavior**
Changing the **Site** field must update the information of the Vanity URL correctly.

**Additional context**
The only workaround is to re-create the Vanity URL from scratch. However, this is not a feasible solution for the customer. We need to look at the possibility of creating a SQL query that could update the incorrect and fix this automatically.

In the customer instance, running 5.3.8.6, I was able to confirm that the "more technical" workaround for fixing a Vanity URL that is not working is the following:

- Look for the Vanity URL in the `identifier` table, and update the `host_inode` column to point to either the right site or `SYSTEM_HOST` if you need it to apply to all Sites.
- Flush the region called `Vanity URL Cache`.
- Test the Vanity URL once more. You can use CURL, like this:
```
curl -sk --head -H "Host: www.yoursite.com" https://your-dotcms-instance.com/your-vanity-url
```

### PR DESCRIPTION
**Pull request by freddyucv**
Flush the Old Host cache when You Update a VanityURL

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java [hunk 1]
```diff
@@ -1941,6 +1942,19 @@ protected Contentlet save(final Contentlet contentlet, final String existingInod
         final Contentlet toReturn = findInDb(inode).orElseThrow(()->
                 new DotStateException(String.format("Contentlet with inode '%s' not found in DB", inode)));
 
+        final List<ContentletVersionInfo> contentletVersionInfos = APILocator.getVersionableAPI()
```
[the remark this answers is not in the document]
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java]: done https://github.com/dotCMS/core/pull/21822/commits/c1ca4bc1b0933dc60bed2bd5e6864480bad64128

[the remark this answers is not in the document]
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java]: done https://github.com/dotCMS/core/pull/21822/commits/c1ca4bc1b0933dc60bed2bd5e6864480bad64128

**Review by jdotcms — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java [hunk 2]
```diff
@@ -1955,6 +1962,22 @@ protected Contentlet save(final Contentlet contentlet, final String existingInod
         return toReturn;
     }
 
+    private void setOlhHostIDProperty(final Contentlet contentlet)
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java:1965]: fix method name

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java [hunk 3]
```diff
@@ -1941,6 +1942,12 @@ protected Contentlet save(final Contentlet contentlet, final String existingInod
         final Contentlet toReturn = findInDb(inode).orElseThrow(()->
                 new DotStateException(String.format("Contentlet with inode '%s' not found in DB", inode)));
 
+
+
+        if (contentlet.isVanityUrl()) {
+            setOlhHostIDProperty(toReturn);
```
[the remark this answers is not in the document]
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java]: done https://github.com/dotCMS/core/pull/21822/commits/35d9b51d6ab5b0397379892b861835eda23e9bcd

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java (hunk 2 above)
Comment (freddyucv) [reply dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java]: done https://github.com/dotCMS/core/pull/21822/commits/35d9b51d6ab5b0397379892b861835eda23e9bcd

**Review by dsilvam — APPROVED** [review]

**Pull request merged**